filter.c

```diff
@@ -26,31 +49,33 @@ int filter(struct xdp_md *ctx)
 	struct iphdr *ip = data + sizeof(struct ethhdr);
 
 	if (ip + 1 > (struct iphdr *)data_end)
-		goto pass;
+		return XDP_DROP;
 
 	if (ip->protocol != IPPROTO_TCP)
-		goto pass;
+		goto end;
 
 	struct tcphdr *tcp = (void *)ip + (ip->ihl * 4);
 	if (tcp + 1 > (struct tcphdr *)data_end)
-		goto pass;
+		goto end;
 
 	__u16 dport = bpf_ntohs(tcp->dest);
 	if ((dport != 80 && dport != 443) || !tcp->psh)
-		goto pass;
+		goto end;
 
 	void *payload = (void *)tcp + (tcp->doff * 4);
 	if (payload > data_end)
-		goto pass;
+		goto end;
 
 	arg.dport = dport;
 	arg.offset = (__u16)(payload - data);
 	arg.length = (__u16)(data_end - payload);
 
 	int action = bpf_luaxdp_run(runtime, sizeof(runtime), ctx, &arg, sizeof(arg));
-	return action < 0 ? XDP_PASS : action;
-pass:
-	return XDP_PASS;
+	if (action >= 0 && action != XDP_PASS) {
```

Comment:
what's the desired effect here? allowing the user to DROP and TX? I found this a bit confusing but perhaps it's only me..

Reply:
It means the `XDP_PASS`, returned by `bpf_luaxdp_run`, will always have the semantics of what I initially called the "default_action"; when in router-mode, `XDP_PASS` would be called, otherwise, `XDP_TX`.

The idea is that we don't want to move this decision to inside the `bpf_luaxdp_run` function, so I assumed `XDP_PASS` means `accept`. 

To state it concisely, the line means: if error or accept, accept; otherwise, forward the returned action.

Reply:
I think it's a bit confusing having the script returning `PASS` and changing it to `TX`.. I would rather let the script to return the proper verdict.. based on the config file perhaps..